iothub/service/src/IotHubCredential.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Azure.Amqp;
+using Microsoft.Azure.Devices.Common;
+
+namespace Microsoft.Azure.Devices
+{
+    /// <summary>
+    /// The properties required for authentication to IoT hub that are independent of the authentication type.
+    /// </summary>
+    internal abstract class IotHubCredential
+        : IAuthorizationHeaderProvider, ICbsTokenProvider
+    {
+        private const string HostNameSeparator = ".";
+
+        public string IotHubName { get; protected set; }
+
+        public string HostName { get; protected set; }
+
+        public Uri HttpsEndpoint { get; protected set; }
+
+        public Uri AmqpEndpoint { get; protected set; }
+
+        public abstract string GetAuthorizationHeader();
+
+        public abstract Task<CbsToken> GetTokenAsync(Uri namespaceAddress, string appliesTo, string[] requiredClaims);
+
+        // Azure.Core (used in IotHubTokenCredential) is not available in NET451.
+        // So we need this constructor for the build to pass.
```

Comment:
Interesting! 
Q - Is there an advantage to the adding this to the base class instead of adding it to IotHubTokenCredential directly?

Reply:
this = the no-args protected constructor

Reply:
Sort of. The problem is if I add it to the IotHubTokenCredential class, I need to pass something to the base constructor for 'connectionString'. This will lead to a nullref exception. Infact it reminds me I need to add a constructor that throws an exception in the IotHubTokenCredential class (which I missed somehow). The base class constructor always gets executed before derived so I cannot avoid the nullref exception.